include/irr/asset/CTexturePacker.h

```diff
@@ -0,0 +1,107 @@
+#ifndef __IRR_C_TEXTURE_PACKER_H_INCLUDED__
+#define __IRR_C_TEXTURE_PACKER_H_INCLUDED__
+
+#include "irr/asset/format/EFormat.h"
+#include "irr/asset/ICPUImage.h"
+#include "irr/core/alloc/GeneralpurposeAddressAllocator.h"
+
+namespace irr {
+namespace asset
+{
+
+class CTexturePacker
+{
+    using addr_alctr_t = core::GeneralpurposeAddressAllocator<uint32_t>;
+
+public:
+    struct STexOffset
+    {
+        uint32_t x, y;
+    };
+
+    CTexturePacker(E_FORMAT _format, uint32_t _szxy = 1u<<14, uint32_t _pgSzxy = 256u) :
+        m_pgSzxy(_pgSzxy),
+        m_alctrReservedSpace(core::make_refctd_dynamic_array<decltype(m_alctrReservedSpace)>(
+            addr_alctr_t::reserved_size(_szxy / _pgSzxy, _szxy / _pgSzxy, 1ull))
+        ),
+        m_addrAlctr(reinterpret_cast<void*>(m_alctrReservedSpace->data()), 0u, 0u, _szxy/_pgSzxy, _szxy,_pgSzxy, 1u)
+    {
+        assert(core::isPoT(_szxy));
+        assert(core::isPoT(_pgSzxy));
+
+        ICPUImage::SCreationParams params;
+        params.extent = {_szxy,_szxy,1u};
+        params.format = _format;
+        params.arrayLayers = 1u;
+        params.mipLevels = 1u;
+        params.type = IImage::ET_2D;
+        params.samples = IImage::ESCF_1_BIT;
+        params.flags = static_cast<IImage::E_CREATE_FLAGS>(0);
+
+        m_megaimg = ICPUImage::create(std::move(params));
+        auto regions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<IImage::SBufferCopy>>(1ull);
+        auto& region = regions->front();
+        region.imageSubresource.mipLevel = 0u;
+        region.imageSubresource.baseArrayLayer = 0u;
+        region.imageSubresource.layerCount = 1u;
+        region.bufferOffset = 0u;
+        region.bufferRowLength = _szxy;
+        region.bufferImageHeight = 0u; //tightly packed
+        region.imageOffset = {0u, 0u, 0u};
+        region.imageExtent = {_szxy,_szxy,1u};
+        auto buffer = core::make_smart_refctd_ptr<ICPUBuffer>(getTexelOrBlockBytesize(_format)*_szxy*_szxy);
+        m_megaimg->setBufferAndRegions(std::move(buffer), regions);
+    }
+
+    STexOffset pack(const ICPUImage* _img)
+    {
+        assert(m_megaimg->getCreationParameters().format==_img->getCreationParameters().format);
+
+        auto extent = _img->getCreationParameters().extent;
+        uint32_t rnd = core::roundUpToPoT(std::max(extent.width,extent.height))/m_pgSzxy;
+        uint32_t pgCnt = rnd*rnd;
```

Comment:
This will get you better packing
```cpp
core::vectorSIMDu32 extentInPages = (extent+pageSize-1u)/pageSize;
uint32_t mortonEnd = core::xyToMorton(extentInPages-core::vectorSIMDu32(1u))+1u; // one past last pixel
uint32_t pgCnt = core::roundUpToPoT(mortonEnd); // will make sure we have a full Z-square or two Z side by side
```